File: dataflow/web/static/metrics.js
// Metrics panel rendering and export.

import { STATES, STATE_COLORS } from './graph.js';

const el = (tag, attrs = {}, children = []) => {
  const node = document.createElement(tag);
  Object.entries(attrs).forEach(([key, value]) => {
    if (key === 'class') node.className = value;
    else if (key === 'text') node.textContent = value;
    else if (value !== null && value !== undefined) node.setAttribute(key, value);
  });
  children.forEach((child) => node.append(child));
  return node;
};

const compact = new Intl.NumberFormat(undefined, {
  notation: 'compact', maximumFractionDigits: 2,
});

function number(value, digits = 3) {
  if (!Number.isFinite(value)) return '—';
  if (value !== 0 && Math.abs(value) >= 100000) return compact.format(value);
  const text = value.toFixed(digits);
  // Trim trailing zeros only past a decimal point: "1660" must not become "166".
  const trimmed = text.includes('.')
    ? text.replace(/0+$/, '').replace(/\.$/, '')
    : text;
  return trimmed || '0';
}

/**
 * A ratio that may be genuinely infinite (sleeping with zero idle cycles to
 * divide by). JSON has no literal for infinity, so the API sends the string
 * "Infinity" / "-Infinity" instead of a bare numeric token the browser's
 * JSON.parse would refuse to parse -- render that sentinel as an actual
 * infinity symbol rather than running it through `number()`, which would
 * just read it as not-a-number and fall back to a generic "no value" dash.
 */
function ratioText(value, digits = 3) {
  if (value === 'Infinity') return '∞';
  if (value === '-Infinity') return '-∞';
  return number(value, digits);
}

function stat(label, value, unit) {
  const box = el('div', { class: 'stat' });
  const dd = el('dd', { text: value });
  if (unit) dd.append(el('small', { text: ` ${unit}` }));
  box.append(el('dt', { text: label }), dd);
  return box;
}

export function download(filename, content, type = 'application/json') {
  const url = URL.createObjectURL(new Blob([content], { type }));
  const anchor = Object.assign(document.createElement('a'), { href: url, download: filename });
  document.body.append(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}

export function toCSV(metrics) {
  const header = ['id', 'name', 'kind', 'firings', 'wakeups', 'utilization',
    'sleep_idle_ratio', 'wakeup_firing_ratio', 'energy', 'avg_power',
    'tokens_in', 'tokens_out', ...STATES];
  const rows = metrics.actors.map((actor) => [
    actor.id, actor.name, actor.kind, actor.firings, actor.wakeups, actor.utilization,
    actor.sleep_idle_ratio, actor.wakeup_firing_ratio,
    actor.energy, actor.avg_power, actor.tokens_in, actor.tokens_out,
    ...STATES.map((state) => actor.state_cycles[state]),
  ]);
  return [header, ...rows].map((row) => row.join(',')).join('\n');
}

export function renderMetrics(container, metrics, { graphName }) {
  const sinks = metrics.actors.filter((actor) => actor.kind === 'sink');

  const stats = el('dl', { class: 'stat-grid' });
  stats.append(
    stat('Avg power', number(metrics.avg_power, 4)),
    stat('Energy', number(metrics.energy, 1)),
    stat(sinks.length > 1 ? 'Throughput (sum)' : 'Throughput',
      number(metrics.throughput, 5), 'tok/cyc'),
    stat('Tokens out', number(metrics.tokens_consumed, 0)),
  );

  const parts = [stats];

  if (metrics.deadlock_cycle !== null && metrics.deadlock_cycle !== undefined) {
    parts.push(el('p', {
      class: 'run-status is-error',
      text: `Deadlocked at cycle ${metrics.deadlock_cycle}: no actor could make progress. `
        + 'Figures still cover the full window.',
    }));
  }

  parts.push(el('p', {
    class: 'hint',
    text: `${metrics.cycles.toLocaleString()} cycles on the ${metrics.engine} engine in `
      + `${metrics.wall_time.toFixed(3)}s · source/sink energy ${number(metrics.special_energy, 1)} `
      + '(excluded above)',
  }));

  if (sinks.length) {
    const perSink = el('div', { class: 'section' });
    perSink.append(el('span', { class: 'panel-label', text: 'Per sink' }));
    sinks.forEach((sink) => {
      const rate = (metrics.per_sink_throughput || {})[sink.id] ?? 0;
      const row = el('div', { class: 'sink-row' });
      row.append(
        el('b', { text: sink.name }),
        el('span', {
          class: 'sink-figures',
          text: `${number(rate, 5)} tok/cyc · ${number(sink.tokens_in, 0)} tokens`,
        }),
      );
      perSink.append(row);
    });
    if (sinks.length > 1) {
      perSink.append(el('p', {
        class: 'hint',
        text: 'The headline throughput is the sum of these — a raw token count, so '
          + 'it only means something when every sink counts the same kind of token.',
      }));
    }
    parts.push(perSink);
  }

  const list = el('div', { class: 'section' });
  list.append(el('span', { class: 'panel-label', text: 'Per actor' }));

  metrics.actors.forEach((actor) => {
    const row = el('div', { class: 'actor-metric' });
    const head = el('div', { class: 'actor-metric-head' });
    head.append(
      el('b', { text: actor.name }),
      el('span', { text: `E ${number(actor.energy, 1)}` }),
    );
    const sub = el('span', {
      class: 'actor-metric-sub',
      text: `${actor.kind_label || actor.kind} · ${actor.firings} firings · `
        + `${(actor.utilization * 100).toFixed(1)}% busy`,
    });
    const bar = el('div', {
      class: 'occupancy',
      title: STATES.map((state) => `${state}: ${actor.state_cycles[state]}`).join('\n'),
    });
    const total = STATES.reduce((sum, state) => sum + actor.state_cycles[state], 0) || 1;
    STATES.forEach((state) => {
      const share = (actor.state_cycles[state] / total) * 100;
      if (share > 0) {
        bar.append(el('i', {
          style: `width:${share}%;background:${STATE_COLORS[state]}`,
        }));
      }
    });
    const ratios = el('span', {
      class: 'actor-metric-ratios',
      title: 'Sleeping cycles per idle cycle, and wakeup transitions per firing',
      text: `sleep/idle ${ratioText(actor.sleep_idle_ratio, 3)} · `
        + `wakeups/firings ${number(actor.wakeup_firing_ratio, 3)}`,
    });
    row.append(head, sub, bar, ratios);
    list.append(row);
  });
  parts.push(list);

  const actions = el('div', { class: 'metric-actions' });
  const csvBtn = el('button', { class: 'btn', text: 'Export CSV' });
  csvBtn.addEventListener('click', () => download(`${graphName}-metrics.csv`, toCSV(metrics), 'text/csv'));
  const jsonBtn = el('button', { class: 'btn', text: 'Export JSON' });
  jsonBtn.addEventListener('click', () => download(`${graphName}-metrics.json`, JSON.stringify(metrics, null, 2)));
  actions.append(csvBtn, jsonBtn);
  parts.push(actions);

  container.replaceChildren(...parts);
}

File: dataflow/sim/metrics.py
"""Throughput, power and energy figures derived from a finished run."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ..model.actor import ActorKind, ActorState
from .core import SimulationCore


@dataclass
class ActorMetrics:
    id: str
    name: str
    kind: str
    kind_label: str
    firings: int
    energy: float
    avg_power: float
    state_cycles: dict[str, int]
    tokens_in: int
    tokens_out: int
    utilization: float  # fraction of cycles spent executing

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "name": self.name,
            "kind": self.kind,
            "kind_label": self.kind_label,
            "firings": self.firings,
            "energy": self.energy,
            "avg_power": self.avg_power,
            "state_cycles": self.state_cycles,
            "tokens_in": self.tokens_in,
            "tokens_out": self.tokens_out,
            "utilization": self.utilization,
        }


@dataclass
class RunMetrics:
    """Aggregate results.

    ``energy``/``avg_power`` cover the compute actors only: sources and sinks are
    test-bench infrastructure and are reported separately in ``special_energy``.
    """

    cycles: int
    engine: str
    throughput: float  # tokens per cycle, summed over sinks
    tokens_consumed: int
    tokens_produced: int
    energy: float
    avg_power: float
    special_energy: float
    deadlock_cycle: int | None
    actors: list[ActorMetrics] = field(default_factory=list)
    per_sink_throughput: dict[str, float] = field(default_factory=dict)
    wall_time: float = 0.0

    def to_dict(self) -> dict[str, Any]:
        return {
            "cycles": self.cycles,
            "engine": self.engine,
            "throughput": self.throughput,
            "tokens_consumed": self.tokens_consumed,
            "tokens_produced": self.tokens_produced,
            "energy": self.energy,
            "avg_power": self.avg_power,
            "special_energy": self.special_energy,
            "deadlock_cycle": self.deadlock_cycle,
            "actors": [a.to_dict() for a in self.actors],
            "per_sink_throughput": self.per_sink_throughput,
            "wall_time": self.wall_time,
        }

    def summary(self) -> str:
        lines = [
            f"engine            {self.engine}",
            f"cycles            {self.cycles}",
            f"throughput        {self.throughput:.6f} tokens/cycle",
            f"tokens produced   {self.tokens_produced}",
            f"tokens consumed   {self.tokens_consumed}",
            f"energy (compute)  {self.energy:.3f}",
            f"avg power         {self.avg_power:.6f}",
            f"energy (src/snk)  {self.special_energy:.3f}",
        ]
        if self.deadlock_cycle is not None:
            lines.append(f"DEADLOCK at cycle {self.deadlock_cycle}")
        lines.append("")
        header = f"{'actor':<14}{'kind':<13}{'firings':>9}{'util':>8}{'energy':>12}"
        lines.append(header)
        lines.append("-" * len(header))
        for actor in self.actors:
            lines.append(
                f"{actor.name[:13]:<14}{actor.kind_label[:12]:<13}{actor.firings:>9}"
                f"{actor.utilization:>7.1%}{actor.energy:>12.3f}"
            )
        return "\n".join(lines)


def collect(core: SimulationCore, cycles: int, engine: str, wall_time: float = 0.0)\
        -> RunMetrics:
    actors: list[ActorMetrics] = []
    compute_energy = 0.0
    special_energy = 0.0
    consumed = 0
    produced = 0
    per_sink: dict[str, float] = {}

    for rt in core.order:
        state_cycles = {s.value: rt.state_cycles[s] for s in ActorState}
        # Integrate energy once, in a fixed state order, so the result does not
        # depend on how the engine chunked time.
        energy = sum(rt.actor.power.of(s) * rt.state_cycles[s] for s in ActorState)
        metrics = ActorMetrics(
            id=rt.id,
            name=rt.actor.name,
            kind=rt.actor.kind.value,
            kind_label=rt.actor.kind.label,
            firings=rt.firings,
            energy=energy,
            avg_power=energy / cycles if cycles else 0.0,
            state_cycles=state_cycles,
            tokens_in=rt.tokens_in,
            tokens_out=rt.tokens_out,
            utilization=(rt.state_cycles[ActorState.EXECUTING] / cycles)
            if cycles else 0.0,
        )
        actors.append(metrics)
        if rt.actor.is_special:
            special_energy += energy
        else:
            compute_energy += energy
        if rt.actor.kind is ActorKind.SINK:
            consumed += rt.tokens_in
            per_sink[rt.id] = rt.tokens_in / cycles if cycles else 0.0
        if rt.actor.kind is ActorKind.SOURCE:
            produced += rt.tokens_out

    return RunMetrics(
        cycles=cycles,
        engine=engine,
        throughput=consumed / cycles if cycles else 0.0,
        tokens_consumed=consumed,
        tokens_produced=produced,
        energy=compute_energy,
        avg_power=compute_energy / cycles if cycles else 0.0,
        special_energy=special_energy,
        deadlock_cycle=core.deadlocked_at,
        actors=actors,
        per_sink_throughput=per_sink,
        wall_time=wall_time,
    )
